include/seqan3/index/fm_index.hpp

```diff
@@ -0,0 +1,336 @@
+// ============================================================================
+//                 SeqAn - The Library for Sequence Analysis
+// ============================================================================
+//
+// Copyright (c) 2006-2018, Knut Reinert & Freie Universitaet Berlin
+// Copyright (c) 2016-2018, Knut Reinert & MPI Molekulare Genetik
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+//     * Redistributions of source code must retain the above copyright
+//       notice, this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above copyright
+//       notice, this list of conditions and the following disclaimer in the
+//       documentation and/or other materials provided with the distribution.
+//     * Neither the name of Knut Reinert or the FU Berlin nor the names of
+//       its contributors may be used to endorse or promote products derived
+//       from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL KNUT REINERT OR THE FU BERLIN BE LIABLE
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+// LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+// OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+// DAMAGE.
+//
+// ============================================================================
+
+/*!\file
+ * \author Christopher Pockrandt <christopher.pockrandt AT fu-berlin.de>
+ * \brief Contains the unidirectional seqan3::fm_index.
+ */
+
+ #include <sdsl/suffix_trees.hpp>
+
+ #include <range/v3/view/reverse.hpp>
+ #include <range/v3/view/transform.hpp>
+
+#include <seqan3/index/concept.hpp>
+#include <seqan3/index/fm_index_iterator.hpp>
+#include <seqan3/core/metafunction/range.hpp>
+
+namespace seqan3
+{
+
+/*!\addtogroup index
+ * \{
+ */
+
+// TODO(h-2): what is ingroup for (vs. addtogroup)?
```

Comment:
I'm not Hannes, but still:

`addtogroup` is like `defgroup`, but instead of throwing a warning if you define one group multiple times, `addtogroup` will merge the documentations of the groups.
So in your example, the `index` group will always work, whether or not it has been defined in another file already.

`ingroup` assigns a block to a group.

By using `ingroup fm_index` within an `addtogroup index` section, you say `fm_index` is a subgroup of `index`.

[Reference](https://www.stack.nl/~dimitri/doxygen/manual/grouping.html)

Reply:
Would make sense to split it up this way.
You can still name the subgroup `fm_index`, since we use singular for the group names (.e.g. Alphabet).

Reply:
I think that people might not be aware of a bidirectional index hiding in the subgroup `fm_index` when there is a unidirectional index with the exact same name

Reply:
We could do

- `group index`
    - `subgroup fm_index`
    - `subgroup 2fm_index`
    - `subgroup k_mer_index`

Then it would be obvious that there is a 2fm.


But still, if you have a `fm_index` module, users would have to look at the docs and in the `fm_index` docs you can also find the 2fm index.

Kinda like in the alphabet module: http://docs.seqan.de/seqan/3.0.0-master-dev/group__alphabet.html
And if you go down one level: http://docs.seqan.de/seqan/3.0.0-master-dev/group__nucleotide.html
And I think we can make it pretty obvious that there are multiple fm indices there, just like in the nuleotides module.

Reply:
Sounds like a good idea!

As discussed with Hannes, we will name it `bi_fm_index`. `fm2_index` rather sounds like version 2 instead of bidirectional (and `2fm_index` is not possible since the language doesn't allow us to start names with a digit)